Expose in-place geometry POSITION buffer updates (VertexBuffer::setBufferAt) via thermion_dart

JesperBellenbaum opened the pull request:
This PR adds a minimal API to update an existing geometry asset’s vertex positions in-place, instead of destroying/recreating geometry each update. The intent is to enable efficient dynamic meshes.


  - Native C API:
      - SceneAsset_updatePositionBuffer(TEngine*, TSceneAsset*, float* positions, uint32_t numFloats) updates buffer slot 0 (POSITION) on geometry assets after validating type and vertexCount * 3.
      - Memory is copied into heap-owned storage and freed via the BufferDescriptor callback.
  - Render-thread proxy:
      - SceneAsset_updatePositionBufferRenderThread(...) enqueues the update on the render thread and returns a success/failure boolean via callback.
  - Dart API:
      - FilamentApp.updateGeometryPositions(ThermionAsset asset, Float32List positions)
      - FFIFilamentApp implementation forwards to the render-thread API and throws if the native update fails.

  Notes

  - This is intentionally “positions-only” and targets the existing geometry layout where POSITION is in buffer slot 0.
  - I’m not a C++ expert; a proper C++/Filament review would be appreciated, especially around render-thread usage, memory/lifetime, and callback conventions.

nmfisher commented:
Thanks for this. I recently added Dart wrappers for VertexBuffer/IndexBuffer, so I'm wondering if a better way to handle this would be to return those instances from GeometrySceneAsset and use the existing Dart methods, e.g.

TSceneAsset.cpp
```
// reserve primitiveIndex for future use, for now geometry assets only support a single primitive so it's not actually used
EMSCRIPTEN_KEEPALIVE TVertexBuffer *SceneAsset_getVertexBuffer(TSceneAsset *tSceneAsset, int primitiveIndex) {

            if (asset->getType() == SceneAsset::SceneAssetType::Geometry) {
                auto geometrySceneAsset = reinterpret_cast<GeometrySceneAsset *>(asset);
                auto *vertexBuffer = geometrySceneAsset->getVertexBuffer();
                return reinterpret_cast<TVertexBuffer *>(vertexBuffer);
            }
            return nullptr;
}
```

ffi_asset.dart
```
VertexBuffer? getVertexBuffer() {
     final vertexBufferPtr = SceneAsset_getVertexBuffer(asset, 0);
     return FFIVertexBuffer(vertexBufferPtr);
}
```

Then in your application code:

```
final vb = asset.getVertexBuffer();
final newVertices = Float32List(await vb.getVertexCount() * 3);
newVertices.setRange(0, newVertices.length, [your values]):
if(vb != null) {
    await vb.setBufferAt(0, newVertices)
}
```

This avoids duplicating the logic for allocating/freeing memory and also remains a bit more flexible for future asset types. It still assumes you know what the correct buffer index is (e.g. 0 for POSITION) but in future we can probably expose/combine RenderableManager::getEnabledAttributesAt to let the user correctly identify the position of arbitrary attributes. 

Can you see any problems with this approach?

JesperBellenbaum committed:
feat: expose geometry VertexBuffer from ThermionAsset

JesperBellenbaum commented:
Ah good point, wasn't aware of these wrappers! 

 - I reset the previous implementation and switched to the approach you suggested: expose the underlying VertexBuffer from SceneAsset and reuse the existing Dart VertexBuffer API (getVertexCount, setBufferAt, destroy).
  - Added native SceneAsset_getVertexBuffer(TSceneAsset*, int primitiveIndex) and wired it through to Dart as ThermionAsset.getVertexBuffer({primitiveIndex = 0}).

nmfisher commented:
Looks good, thanks for the contribution. We should probably add a test & expose the index buffer too, but I'll do those separately as I can't push to this PR.

Also I see that you developed the flutter_gaussian_splatter package, which is great work.  I've been dabbling with 3D splats recently, so would be interested in contributing if there's anything similar.

JesperBellenbaum commented:
Thanks! PRs are very welcome there :)
